pilot/pkg/xds/eds.go

```diff
@@ -249,6 +240,34 @@ func (s *DiscoveryServer) deleteService(cluster, serviceName, namespace string)
 	}
 }
 
+// UpdateServiceAccount updates the service endpoints' sa when service/endpoint event happens.
+func (s *DiscoveryServer) UpdateServiceAccount(shards *EndpointShards, clusterID, serviceName string, endpoints []*model.IstioEndpoint) bool {
+	oldServiceAccount := shards.ServiceAccounts
+	shards.mutex.Lock()
+	defer shards.mutex.Unlock()
+
+	// Check if ServiceAccounts have changed. We should do a full push if they have changed.
+	serviceAccounts := sets.Set{}
+	for cluster, epShards := range shards.Shards {
+		if cluster == clusterID {
+			epShards = endpoints
+		}
+		for _, ep := range epShards {
+			if ep.ServiceAccount != "" {
+				serviceAccounts.Insert(ep.ServiceAccount)
+			}
+		}
+	}
+	shards.ServiceAccounts = serviceAccounts
```

Comment:
>Do not we have update only it is changed?
Strictly, we should.

>Can we take a read lock and get shards - then compute service accounts and if they change then only take a write lock and update? Does it make any difference?

It is not worth so much trouble, and if so it may introduce in-consistent between read/write interval.